docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+  - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.
+
+##### OS Update Failures
+
+- **Before rebooting for update** - Update cannot start if the machine image version is unavailable, network issues prevent the image pull, or disk storage is insufficient (although proper garbage collection should prevent this); classified as retriable or non-retriable.
+- **Failed to boot the new OS** - If the node fails to boot into the new OS version and reverts to the previous version, `gardener-node-agent` will annotate it to mark an update attempt, but the root cause may remain unlogged as filesystems mount later in the booting process; classified as non-retriable.
+- **Total failure** - In cases where the system fails to boot into either of the OS versions, the update attempt will timeout and be marked as failed; classified as non-retriable.
+
+![FailureStrategy](assets/inplace-failurestrategy.png "InPlace Failure State Diagram")
+
+#### Roles in Failure Handling
+
+**`gardener-node-agent`**
+`gardener-node-agent` is responsible for initiating updates, with specific timeouts set for Kubernetes and machine image updates. When an error occurs it will be classified as follows:
+
+- **Retriable Error**: Temporary errors, such as network interruptions, prompt `gardener-node-agent` to retry the update after a delay. Retriable error will also get classified as non-retriable error after a certain number of retries and `gardener-node-agent` labels the node with an `update-failed` label.
+- **Non-Retriable Error**: For critical issues, such as incompatible OS versions, `gardener-node-agent` labels the node with an `update-failed` label.
+
+**`machine-controller-manager`**
+`machine-controller-manager` prepares nodes for updates by adding the `ready-to-update` label and waits for `gardener-node-agent` to perform the update. `machine-controller-manager` monitors for status labels applied by `gardener-node-agent`:
+
+- **Timeout Handling**: After the `machine-controller-manager` adds the `ready-for-update` label to a node, it waits for the `gardener-node-agent` to complete the update within the timeout configured in `machine-controller-manager`. If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay. The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update. If the update is unsuccessful after a certain number of retries, `machine-controller-manager` will add the `update-failed` label to the node.
```

Comment:
How do we ensure that this will not be a race? If both `gardener-node-agent` and `machine-controller-manager` may set the update label what happens if we run close to the timeout and they overwrite each others value? Any thoughts?

Reply:
We are using labels of the form `node.machine.sapcloud.io/ready-for-update: true` and `node.machine.sapcloud.io/update-successful: true`. `gardener-node-agent` and MCM will be working on different labels and they should not be writing to the same label at any point.